cpp/src/packed/writer.cpp

```diff
@@ -35,40 +37,43 @@ PackedRecordBatchWriter::PackedRecordBatchWriter(size_t memory_limit,
       file_path_(file_path),
       props_(props),
       splitter_({}),
-      current_memory_usage_(0) {}
+      current_memory_usage_(0),
+      size_split_done_(false) {}
 
-Status PackedRecordBatchWriter::Init(const std::shared_ptr<arrow::RecordBatch>& record) {
-  // split first batch into column groups
-  std::vector<ColumnGroup> groups = SizeBasedSplitter(record->num_columns()).Split(record);
+Status PackedRecordBatchWriter::Write(const std::shared_ptr<arrow::RecordBatch>& record) {
+  if (!size_split_done_) {
+    size_t next_batch_size = GetRecordBatchMemorySize(record);
+    if (current_memory_usage_ + next_batch_size < memory_limit_) {
```

Comment:
I believe the upper bound of memory in the probing phase should be `memory_limit_ / 2`, or else the total use of memory could exceed `memory_limit_` when the records are actually written to the buffer.